contributors/design-proposals/pod-preemption.md

```diff
@@ -0,0 +1,188 @@
+-  Define how priority and other metrics affect preemption.
+-  Define scenarios under which a pod may get preempted.
+-  Define the interaction between scheduler preemption and Kubelet evictions.
+-  Define mechanics of preemption.
+
+## Non-Goals
+
+-  How **eviction** works in Kubernetes. (Please see [Background](#background) for the definition of "eviction".)
+-  How quota management works in Kubernetes.
+
+# Background
+
+Running various types of workloads with different priorities is a common practice in medium and large clusters to achieve higher resource utilization. In such scenarios, the amount of workload can be larger than what the total resources of the cluster can handle. If so, the cluster chooses the most important workloads and runs them. The importance of workloads are specified by a combination of [priority](https://github.com/bsalamat/community/blob/564ebff843532faf5dcb06a7e50b0db5c5b501cf/contributors/design-proposals/pod-priority-api.md), QoS, or other cluster-specific metrics. The potential to have more work than what cluster resources can handle is called "overcommitment". Overcommitment is very common in on-prem clusters where the number of nodes is fixed, but it can similarly happen in cloud as cloud customers may choose to run their clusters overcommitted/overloaded at times in order to save money. For example, a cloud customer may choose to run at most 100 nodes, knowing that all of their critical workloads fit on 100 nodes and if there is more work, they won't be critical and can wait until cluster load decreases.
+
+## Terminology
+
+When a new pod has certain scheduling requirements that makes it infeasible on any node in the cluster, scheduler may choose to kill lower priority pods to satisfy the scheduling requirements of the new pod. We call this operation "**preemption**". Preemption is distinguished from "[**eviction**](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#eviction-policy)" where kubelet kills a pod on a node because that particular node is running out of resources.
+
+# Overview
+
+This document describes how preemption in Kubernetes works. Preemption is the action taken when an important pod requires resources or conditions which are not available in the cluster. So, one or more pods need to be killed to make room for the important pod. 
+
+# Detailed Design
+
+## Preemption scenario
+
+In this proposal, the only scenario under which a group of pods in Kubernetes may be preempted is when a higher priority pod cannot be scheduled due to various unmet scheduling requirements, such as lack of resources, unsatisfied affinity or anti-affinity rules, etc., and the preemption of the lower priority pods allows the higher priority pod to be scheduled. So, if the preemption of the lower priority pods does not help with scheduling of the higher priority pod, those lower priority pods will keep running and the higher priority pod will stay pending.  
```

Comment:
+1

The one thing that's not covered is if users want to grow the cluster instead of preempting, just to avoid the disruption (i.e. instead of preempting and then growing the cluster because the victim goes pending, avoid the preemption altogether by growing the cluster when the high-priority pod is submitted). Obviously one way to do this is to set all pods at the same priority (i.e. not use the priority/preemption feature) but it would be nice if clusters that use priority/preemption could somehow set the tradeoff. PDB might play a role here -- we could say that if a preemption would violate PDB then we will try to grow the cluster, and only preempt if the cluster has hit the max size the user configured for the cluster autoscaler. I suspect that in the absence of PDB, the current behavior is probably what users want because growing the cluster takes a while, whereas preemption allows the high-priority pod to get scheduled right away (and then if the victim has to wait a while for the new node to spin up, it's not a big deal). Keep in mind that PDB will eventually support preemption rate limits, not just minAvailable/maxUnavailable.